Add Content Helper feature

pauarge opened the pull request:
## Description
This PR introduces a new feature called **Content Helper**, which is the main feature of the [3.4.0 milestone](https://github.com/Parsely/wp-parsely/milestone/34).

### How it works:
- The Content Helper adds a Parse.ly icon in the Gutenberg editor.
- Clicking the icon presents a Gutenberg sidebar, displaying the top performing posts based on the current post's `tag` or `category` (in order of priority). From the sidebar, it is also possible to visit the specific posts, or their respective Parse.ly Dash pages.
- If the current post does not have any tag or category assigned, the Content Helper displays the top posts by the current `author` (WP user).

### Things of note:
- This is an MVP and the first iteration on this feature (possible feature requests should be posted in the repo's issues).
- If the post contains multiple categories/tags, only the first category/tag (as returned by WordPress) is taken into account.
- Currently, only one filter (tag, category, author) is taken into account. That means that the Content Helper doesn’t support queries of `and` type (for example, author + tag).
- Posts are considered as "Top posts" based on their number of views.
- To get the list of posts, we use Parse.ly's `/analytics/posts` API endpoint.
- This PR also contains a related small refactoring to reduce code duplication.

## Motivation and Context
Start implementing tools within the Gutenberg editor that will be providing help and insights to our users, so they can get even more value from their Parse.ly integration.

## How Has This Been Tested?
At this point, manual testing in a local devenv only. Another PR adding tests will be be posted in the coming days. The reasoning behind this was to make the current PR available for review as soon as possible, since the intent is to release it next week if possible.

## Screenshot

![image](https://user-images.githubusercontent.com/23142906/172695623-19310989-f319-4d24-bfe5-dab4fa415926.png)

pauarge committed:
Adding content helper branch

pauarge committed:
Merge branch 'develop' into feature/content-helper

pauarge committed:
Updating requirements

pauarge committed:
Adding placeholder content

pauarge committed:
Moving to cards format

pauarge committed:
Merge branch 'develop' into feature/content-helper

pauarge committed:
Removing duplicate logo

pauarge committed:
Refactoring plugin creation

pauarge committed:
Extracting PostCard component

pauarge committed:
Adding provider

pauarge committed:
Adding provider

pauarge committed:
Add data as dev dependency

pauarge committed:
Adding links

pauarge committed:
Parametric data

pauarge committed:
Merge branch 'develop' into feature/content-helper

pauarge committed:
Updating builds

pauarge committed:
Adding spinner

pauarge committed:
Compacting view

acicovic committed:
Add basic API Secret functions

acicovic committed:
Base_Proxy: Accept API Secret in get_api_url()

acicovic committed:
Cached_Proxy: Update some comments for clarity

acicovic committed:
Add analytics proxy and endpoint

acicovic committed:
Add code that loads the analytics proxy and endpoint

pauarge committed:
Adding Typescript

pauarge committed:
Merge branch 'develop' into feature/content-helper

pauarge committed:
Fix tsconfig

pauarge committed:
Merge branch 'develop' into feature/content-helper

pauarge committed:
Correcting minor linting issues

acicovic committed:
Merge pull request #894 from Parsely/add/analytics-proxy-endpoint-2

Add Analytics Proxy and Endpoint for Content Helper

pauarge committed:
Merge branch 'develop' into feature/content-helper

pauarge committed:
Adjusting card styles

pauarge committed:
Remove duplicate svg

pauarge committed:
Merge branch 'develop' into feature/content-helper

pauarge committed:
Merge branch 'develop' into feature/content-helper

pauarge committed:
Adding fetcher

acicovic committed:
Add /posts to analytics endpoints

acicovic committed:
class-base-proxy.php: Correct 'secret' param name

acicovic committed:
Analytics API Proxy: Fix params not being taken into account

pauarge committed:
Connecting URL to backend

pauarge committed:
Adding basic error handling

pauarge committed:
Merge branch 'develop' into feature/content-helper

pauarge committed:
Adding tag fetching

pauarge committed:
Generating stats url

acicovic committed:
ContentHelperProvider: Add basic querybuilder function

acicovic committed:
ContentHelperProvider: Add built files

pauarge committed:
Merge branch 'develop' into feature/content-helper

acicovic committed:
Content Helper: Rebuild assets

acicovic committed:
Merge branch 'develop' into feature/content-helper

acicovic committed:
Merge branch 'develop' into feature/content-helper

acicovic committed:
Merge branch 'develop' into feature/content-helper

acicovic committed:
Merge branch 'feature/content-helper' of github.com:Parsely/wp-parsely into feature/content-helper

acicovic committed:
Sort devDependencies alphabetically

acicovic committed:
Disable invalid eslint warning about wordpress/core-data/schema

acicovic committed:
Content Helper: Fix non-working suggestion by tag

acicovic committed:
Content Helper: Fix non-working date field

acicovic committed:
Content Helper: Format date according to WP options

acicovic committed:
Content Helper: Display number of views

acicovic committed:
Content Helper: Fix React "unique key prop" error

acicovic committed:
Content Helper: Add error for when Parse.ly API retuns an empty result

acicovic committed:
Merge branch 'develop' into feature/content-helper

acicovic committed:
Content Helper: Add message about the post list's current filter

acicovic committed:
Merge branch 'feature/content-helper' of github.com:Parsely/wp-parsely into feature/content-helper

acicovic committed:
Refactor API proxies to reduce code duplication

acicovic committed:
Fix missing type declaration

acicovic committed:
Content Helper: Rename some ts/tsx files to resemble more to WP standards

acicovic committed:
Content Helper: Add i18n support

acicovic committed:
Content Helper: Add support for unsaved categories ans tags

Previously, we would need to save the post for the Content Helper to intercept categories and tags. This is no longer the case.

GaryJones commented:
What's the goal of this new feature? What do we want editors to know or what actions do we went them to take? How does this help them with their content writing?

acicovic commented:
Hi,

The basic idea behind this is to present writers with similar posts that are the most successful for their website.

This way writers can use successful posts as an inspiration/reference while writing new posts or editing existing ones, and might be able to find patterns of what makes a post successful for their website. For those who haven't thought about this, it opens a new possibility. For writers that have thought of going through this manually, this gives them a shortcut and in-editor way to accomplish this.

acicovic committed:
Merge branch 'develop' into feature/content-helper

acicovic committed:
Rebuild package-lock.json

acicovic committed:
Rebuild package-lock.json

chriszarate commented on src/Endpoints/class-base-api-proxy.php:
```diff
@@ -0,0 +1,144 @@
+			'query' => array(
+				'default'           => array(),
+				'sanitize_callback' => function ( $query ) {
+					// question: how should we sanitize these?
```
I know this is copypasta, but sanitization here should probably map over the array and call `sanitize_text_field` for each member

acicovic replied:
Done.

chriszarate commented on src/Endpoints/class-base-api-proxy.php:
```diff
@@ -0,0 +1,144 @@
+	 */
+	public function permission_callback(): bool {
+		// Unauthenticated.
+		return true;
```
Consider making this an abstract method, so that implementors do not accidentally inherit unauthenticated endpoints.

acicovic replied:
Done.

chriszarate commented on src/Endpoints/class-base-api-proxy.php:
```diff
@@ -0,0 +1,144 @@
+			);
+		}
+
+		if ( true === $require_api_secret && false === $this->parsely->api_secret_is_set() ) {
+			return (object) array(
+				'data'  => array(),
+				'error' => new WP_Error( 400, __( 'A Parse.ly API Secret must be set in site options to use this endpoint', 'wp-parsely' ) ),
+			);
+		}
```
The code in this `if` block can never be reached

acicovic replied:
I've stepped through the code and it is reachable (notably when trying to use the Content Helper but having no secret set in settings). The problem that I noticed with this code (and the block above it) is that it was crashing the WordPress Editor without showing the error message when using the Content Helper.

I've applied a fix for this.

chriszarate replied:
My fault, I was conflating `api_key_is_set` and `api_secret_is_set`. Glad to see I accidentally uncovered a different problem! :)

chriszarate commented on src/Endpoints/class-base-api-proxy.php:
```diff
@@ -0,0 +1,144 @@
+		}
+
+		if ( null !== $param_item ) {
+			$params = $request->get_params()[ $param_item ];
```
use `get_param()`

https://developer.wordpress.org/reference/classes/wp_rest_request/get_param/

acicovic replied:
Done.

chriszarate commented on src/Endpoints/class-base-api-proxy.php:
```diff
@@ -0,0 +1,144 @@
+	 * @param array<string, mixed> $response The response received by the proxy.
+	 * @return array<stdClass> The generated data.
+	 */
+	protected function generate_data( array $response ): array {
```
An abstract method seems useful here, too.

acicovic replied:
Done.

chriszarate commented on src/blocks/content-helper/components/post-card.tsx:
```diff
@@ -0,0 +1,38 @@
+					</p>
+				</CardBody>
+			</Card>
+			<br />
```
Is it possible to style this component and avoid literal line breaks?

acicovic replied:
What's the issue we would be trying to avoid?

chriszarate replied:
Line breaks are a semantic element and are imprecise at applying vertical spacing. It would be better to provide spacing via CSS.

acicovic replied:
Oh, got it. Didn't realize you were talking about the `<br />`. Will look into that.

acicovic replied:
I've tried this but Gutenberg somehow overrides the style and I haven't found a way around this yet. I propose to tackle this in another PR so we can go forward with the release.

chriszarate commented on src/blocks/content-helper/components/post-list.tsx:
```diff
@@ -0,0 +1,55 @@
+				setMessage( result.message );
+				setLoading( false );
+			} )
+			.catch( async ( err: string ) => {
```
Is `err` a `string`? You might want to access `err.message`

acicovic replied:
This seems correct to me.

chriszarate replied:
Unless code somewhere is literally throwing a string (`throw 'a string'`), then `err` is not a string. In the end, via React, you are essentially coercing the error to a string. In other words, something like this is happening:

```
`${ new Error( 'leveraging .toString() method' ) }`
```

It's probably an `Error`, but this is one of those situations where JavaScript is very permissive and you need to do some type narrowing. 

https://kentcdodds.com/blog/get-a-catch-block-error-message-with-typescript

That's a lot of work for a little benefit, and it could be totally fine to coerce to string, but just explaining what's going on under the hood.

acicovic replied:
Thanks for the info, this is something to keep in mind for me.

In this case, [fetchData()](https://github.com/Parsely/wp-parsely/blob/3207db79d69e06711e1e4fc0c1eba9a61e802f34/src/blocks/content-helper/content-helper-provider.ts#L62-L79) returns an array of results or a string error.

Maybe not an ideal solution (?). For the time being I propose to leave as is and work on it in the future if you think it may be worth it?

acicovic replied:
Upon further examination what I said above is partly true, because there's a case where an error is returned.

chriszarate commented on src/blocks/content-helper/components/post-list.tsx:
```diff
@@ -0,0 +1,55 @@
+	const [ message, setMessage ] = useState<string>( null );
+	const [ posts, setPosts ] = useState<SuggestedPost[]>( [] );
+
+	const fetchPosts = async ( retries: number ) => {
```
It's generally better practice to define this function inside `useEffect`:

1. You don't need it to be defined every render.
2. You avoid accidentally calling it outside the effect, which would cause hard-to-debug state issues.

chriszarate commented on src/blocks/content-helper/components/post-list.tsx:
```diff
@@ -0,0 +1,55 @@
+
+	return (
+		<>
+			<p>{ message }</p>
```
Should `message` be rendered even in an error state?

acicovic replied:
Yes. It can contain details about the error.

chriszarate replied:
How, since `setMessage` is only called in the `.then` block / happy path? If an error is set, it seems `message` will always be `null`, which results in an empty `<p/>` being rendered.

It seems we might benefit from an early return in an error case:

```
if ( error ) {
   return <p>{ error }</p>;
}
```

acicovic replied:
The errors we return are strings and they are presented within the message. However, if other errors were to occur what you say is correct. So I've implemented your suggestion.

chriszarate commented on src/blocks/content-helper/content-helper-provider.ts:
```diff
@@ -0,0 +1,118 @@
+		const editor = select( 'core/editor' );
+
+		// Get post's author.
+		const currentPost = editor.getCurrentPost() as Post;
```
Which type is being overridden here?

chriszarate replied:
In general I think these `as` operators should be avoided. Work with the existing return type of these functions, inspect and validate the result, and perform type narrowing to arrive at the expected type.

acicovic replied:
The type returned is a `Record<string, any>`. In that specific case, using `as` simplifies the code and allows to get correct Intellisense quickly. I would go for keeping for the time being as it also should not have side effects as per the [docs](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions):

> Like a type annotation, type assertions are removed by the compiler and won’t affect the runtime behavior of your code.

Having said that, I do agree that manipulating the original type would be the best practice in general.

chriszarate replied:
Sounds good, we can revisit as we get more familiar with Gutenberg dev. :)

chriszarate commented on src/blocks/content-helper/content-helper-provider.ts:
```diff
@@ -0,0 +1,118 @@
+		const author = select( 'core' ).getEntityRecord( 'root', 'user', currentPost.author ) as User;
+
+		// Get post's first category.
+		const categoryId = editor.getEditedPostAttribute( 'categories' ) as Array<number>[0];
```
Isn't `Array<number>[0]` equivalent to `number`?

chriszarate commented on src/blocks/content-helper/content-helper-provider.ts:
```diff
@@ -0,0 +1,118 @@
+		return data.map( ( p ) => {
+			// @ts-ignore
+			p.statsUrl = `${ window.wpParselyContentHelperPrefix }?url=${ p.url }`;
+			return p;
```
- What error is being ignored?
- Encode the URL

```suggestion
			const statsUrl = `${ window.wpParselyContentHelperPrefix }?url=${ window.encodeURIComponent( p.url ) }`;
			return Object.assign( p, { statsUrl } );
```

acicovic replied:
We're getting a "Property 'wpParselyContentHelperPrefix' does not exist on type 'Window & typeof globalThis" error.

chriszarate replied:
Ideally we could import types provided by Parse.ly's JavaScript library. If those don't exist, we can declare our own (alongside others, like for SVG imports).

https://www.jamestharpe.com/typescript-extend-window/

chriszarate commented on src/blocks/content-helper/content-helper.tsx:
```diff
@@ -0,0 +1,31 @@
+ */
+import PostList from './components/post-list';
+// @ts-ignore
+import { ReactComponent as LeafIcon } from '../../img/parsely-logo.svg';
```
https://duncanleung.com/typescript-module-declearation-svg-img-assets/

acicovic committed:
Merge branch 'develop' into feature/content-helper

acicovic committed:
Add sanitization to register_endpoint()

acicovic committed:
Base_API_Proxy: Add abstract methods

acicovic committed:
Base_API_Proxy: Use get_param() instead of get_params() where appropriate

acicovic committed:
Content Helper: Fix error handling

Some errors were not handled correctly by the Content Helper and would make Gutenberg crash. This commit fixes this.

acicovic committed:
Content Helper: Encode URLs

Co-authored-by: Chris Zarate <chris@zarate.org>

acicovic committed:
Content Helper: Re-add ts-ignore to allow build

acicovic committed:
Content Helper PostList: Add fetchPosts() within useEffect()

acicovic committed:
ContentHelperProvider: Refactor code for getting category and tag

acicovic committed:
Import img/parsely-logo.svg correctly

acicovic committed:
Content Helper processData(): Remove ts-ignore

acicovic committed:
Content Helper: Don't display an empty <p> when errors occur

chriszarate reviewed (approved):
Looks good. Did you implement a guard to protect against `undefined` values in the data loader? Or will that come in a later PR?

acicovic commented:
> Looks good. Did you implement a guard to protect against `undefined` values in the data loader? Or will that come in a later PR?

[This](https://github.com/Parsely/wp-parsely/blob/f26ed34cfb74b90f48cb5532f160f5d2ee070681/src/blocks/content-helper/content-helper-provider.ts#L43-L45) acts as the guard, as [buildFetchDataQuery()](https://github.com/Parsely/wp-parsely/blob/f26ed34cfb74b90f48cb5532f160f5d2ee070681/src/blocks/content-helper/content-helper-provider.ts#L88-L96) returns with an error message and `null` query when all of the needed information is `undefined`.
